### PR TITLE
write CSF course name onto blank certificate

### PR DESCRIPTION
**Pull request by davidsbailey**
## Background

Finishes [PLAT-1735](https://codedotorg.atlassian.net/browse/PLAT-1735). Here is the current state of the world:

Screen Shot 2022-06-13 at 9 07 50 AM

Most importantly, when completing a course:
* completing a HOC course will always take you to the studio congrats page
* completing a CSF course will take you to the studio congrats page only if you have the experiment enabled

For more diagrams and next steps, see [congrats page work plan](https://docs.google.com/document/d/1wTObDFZdWS6RzpM3R6wEm04ToY2fM1B5Kb2rUHCNq28/edit).

## Description

The goal of this PR is that when you complete a CSF course with the experiment enabled, you should see the course name printed onto a blank certificate (see 1st screenshot). In order to accomplish that, this PR does the following things:
1. add `default_random_donor` param to `create_course_certificate_image`, and use it to limit random donor selection so that this method never provides a random donor when called from dashboard (e2b162479f9c4d35808f6ee613bee44c582aa624)
1. relax requirements on /certificate_images/ routes so that student name and donor name can be omitted (7ce1869)
1. use custom certificate with course name written onto it (rather than default HOC certificate) for "blank" CSF certificates (https://github.com/code-dot-org/code-dot-org/pull/46829/commits/fb4d4de4fd965bcf42fc8968d87c2a6ec6831477)
1. remove client logic for displaying blank HOC certificate, and use existing server logic instead (https://github.com/code-dot-org/code-dot-org/pull/46829/commits/e6091caa89575617e42c625a7c208d1481cd45e0, https://github.com/code-dot-org/code-dot-org/pull/46829/commits/26c633f2d5de67e40a38e797986f1ee0c9786bbd)

Steps 1-3 are necessary to get blank csf certificates to appear properly on `/certificates/` and `/print_certificates/` pages (see 1st screenshot/video for a refresher). step 4 is also necessary to get blank csf certificates to appear properly on `/congrats/`. 

Also worth noting that step 4 is a "visible" change even for users without the experiment, in the sense that it changes the url used to display the certificate image for HOC tutorials. However, the image itself should be identical (see last screenshot).

## Screenshots

new behavior for finishing CSF congrats page, with experiment:

https://user-images.githubusercontent.com/8001765/173402632-20474bf5-fdb4-49d7-b346-534d0716fa22.mov

after going to code.org/congrats/course1?enableExperiments=studioCertificate (before / after):

Screen Shot 2022-06-13 at 9 17 11 AM

after completing the oceans tutorial, with or without experiment (before / after):

Screen Shot 2022-06-13 at 9 22 13 AM

no image rendered with a [bogus course name](http://localhost-studio.code.org:3000/congrats?i=_1_cba221a6724cf7aaa935202279336e76&s=Ym9ndXM=) (just trying to keep us out of trouble here): 

Screen Shot 2022-06-13 at 5 25 38 PM

## Testing story

* updated and extended unit test coverage on rails controllers and react components
* updated existing UI tests and added new UI / eyes tests

## Follow-up work

* Unfortunately, this PR takes away a usage of images within apps (good) and makes it go back to using the same images in pegasus (bad). I've added an item to the work plan to move the certificate images out of pegasus and into the dashboard asset pipeline (also good).
* This PR has merge conflicts with https://github.com/code-dot-org/code-dot-org/pull/46722. Once both are in, I'll reimplement these commits: [make initialCertificateImageUrl prop required](https://github.com/code-dot-org/code-dot-org/pull/46829/commits/f13af0b8915f1f2c872e72290e32852e13adec3a) and  [extract defaultProps for Congrats](https://github.com/code-dot-org/code-dot-org/pull/46829/commits/5ada958bf17e3799ee87cb45de7cc59470bd02ef)

### CODE REVIEW

**Comment by megcrenshaw** [comment]
Your screencasts are so helpful. I'm having trouble verifying this works locally though:
- Using experiment on Course 1 redirects to studio, but doesn't have Course 1 on it. I also tried inputting my name and tried clicking on the image: 
https://watch.screencastify.com/v/VpihFl3sUqOjOAIyGlBO
- Here I tried clicking "print" and got an Hour of Code certificate: https://watch.screencastify.com/v/pRCctl30bhipCTyafcTx 

I pulled your branch, ran `yarn build` and ran `dashboard-server` ... is there something else I need to do before trying it?

**Review by jamescodeorg — APPROVED** [review]
Changes look good to me (but I didn't try to pull the branch and run it like Meg's doing).  Thanks for continuing to push on this!

**Comment by davidsbailey** [comment]
Apologies, Meg. For some reason, you have to be running the dashboard server on Linux for writing onto certificates to work. I tried to set up an adhoc, but it kept failing due to known issues which infra is working on, so I took the screenshots against a server running on a developer EC2 instance. 

I think that explains the first 2 anomalies, but I'll have to look into the 3rd next week.

**Comment by davidsbailey** [comment]
> * Here I tried clicking "print" and got an Hour of Code certificate: https://watch.screencastify.com/v/pRCctl30bhipCTyafcTx

@megcrenshaw I think this looks like you may not have had the experiment enabled on dashboard. unfortunately, you have to enable it on both dashboard AND pegasus. you can see this happening in the UI tests: https://github.com/code-dot-org/code-dot-org/blob/6b5f4fc836e1b01a4b5409b4c2ab8cc1c34c2c37/dashboard/test/ui/features/hour_of_code/hour_of_code_finish.feature#L61-L65

**Comment by megcrenshaw** [comment]
> @megcrenshaw I think this looks like you may not have had the experiment enabled on dashboard. unfortunately, you have to enable it on both dashboard AND pegasus. you can see this happening in the UI tests:

Ah, thanks! That was the problem. Name doesn't come through locally, but it's not the Hour of Code certificate:

image

File: apps/src/templates/certificates/Certificate.jsx [hunk 1]
```diff
@@ -11,18 +11,6 @@ import SocialShare from './SocialShare';
...
-  hourOfCode: require('@cdo/static/hour_of_code_certificate.jpg'),
-  oceans: require('@cdo/static/oceans_hoc_certificate.png'),
-  mc: require('@cdo/static/MC_Hour_Of_Code_Certificate.png'),
-  minecraft: require('@cdo/static/MC_Hour_Of_Code_Certificate.png'),
-  hero: require('@cdo/static/MC_Hour_Of_Code_Certificate_Hero.png'),
-  aquatic: require('@cdo/static/MC_Hour_Of_Code_Certificate_Aquatic.png'),
-  mee: require('@cdo/static/MC_Hour_Of_Code_Certificate_mee.png'),
-  mee_empathy: require('@cdo/static/MC_Hour_Of_Code_Certificate_mee_empathy.png'),
-  mee_timecraft: require('@cdo/static/MC_Hour_Of_Code_Certificate_mee_timecraft.png')
-};
```
Comment (megcrenshaw) [inline apps/src/templates/certificates/Certificate.jsx:24]: Hurray deleted files!

File: dashboard/test/controllers/certificate_images_controller_test.rb [hunk 2]
```diff
@@ -33,13 +31,20 @@ class CertificateImagesControllerTest < ActionController::TestCase
     assert_includes response.body, 'invalid donor name'
   end
 
-  test 'can show certificate image given bogus course name' do
```
Comment (megcrenshaw) [inline dashboard/test/controllers/certificate_images_controller_test.rb:36]: I like that you added a test to document *undesired* behavior and then revised the test once the functionality was as desired

**Review by megcrenshaw — APPROVED** [review]
Code looks great! I verified what I could locally. Nice work, and thanks for the thorough documentation

**Comment by davidsbailey** [comment]
Thank you both for the thorough review! I tried one last time to start an adhoc, but no luck 🤷

**Pull request merged**